apps/glusterfs/placer_arbiter_test.go

```diff
@@ -0,0 +1,721 @@
+		"10000000",
+		"11111111",
+		"/dev/foobar",
+		21000)
+	dsrc.QuickAdd(
+		"20000000",
+		"22222222",
+		"/dev/foobar",
+		22000)
+	dsrc.QuickAdd(
+		"30000000",
+		"33333333",
+		"/dev/foobar",
+		23000)
+
+	opts := &TestPlacementOpts{
+		brickSize:       50,
+		brickSnapFactor: 0.3,
+		brickOwner:      "asdfasdf",
+		setSize:         3,
+		setCount:        1,
+	}
+
+	abplacer := NewArbiterBrickPlacer()
+	ba, err := abplacer.PlaceAll(dsrc, opts, nil)
+	tests.Assert(t, err == nil, "expected err == nil, got:", err)
+	tests.Assert(t, len(ba.BrickSets) == 1,
+		"expected len(ba.BrickSets) == 1, got:", len(ba.BrickSets))
+	tests.Assert(t, len(ba.BrickSets[0].Bricks) == 3,
+		"expected len(ba.BrickSets[0].Bricks) == 3, got:",
+		len(ba.BrickSets[0].Bricks))
+
+	_, err = abplacer.Replace(dsrc, opts, nil, ba.BrickSets[0], 0)
+	tests.Assert(t, err == ErrNoSpace,
+		"expected err == ErrNoSpace, got:", err)
+}
+
+func TestArbiterBrickPlacerReplaceTooFewArbiter(t *testing.T) {
+	dsrc := NewTestDeviceSource()
+	dsrc.QuickAdd(
+		"10000000",
+		"11111111",
+		"/dev/foobar",
+		21000)
+	dsrc.QuickAdd(
+		"20000000",
+		"22222222",
+		"/dev/foobar",
+		22000)
+	dsrc.QuickAdd(
+		"30000000",
+		"33333333",
+		"/dev/foobar",
+		23000)
+	dsrc.QuickAdd(
+		"40000000",
+		"44444444",
+		"/dev/foobar",
+		24000)
+	dsrc.QuickAdd(
+		"50000000",
+		"a5555555",
+		"/dev/foobar",
+		25000)
+	// we have enough devices for a generic replace but not
+	// when we're limited to certain devices
+
+	opts := &TestPlacementOpts{
+		brickSize:       50,
+		brickSnapFactor: 0.3,
+		brickOwner:      "asdfasdf",
+		setSize:         3,
+		setCount:        1,
+	}
+
+	abplacer := NewArbiterBrickPlacer()
+	abplacer.canHostArbiter = func(d *DeviceEntry) bool {
+		return d.Info.Id[0] == 'a'
+	}
+	abplacer.canHostData = func(d *DeviceEntry) bool {
+		return !abplacer.canHostArbiter(d)
+	}
+
+	ba, err := abplacer.PlaceAll(dsrc, opts, nil)
+	tests.Assert(t, err == nil, "expected err == nil, got:", err)
+	tests.Assert(t, len(ba.BrickSets) == 1,
+		"expected len(ba.BrickSets) == 1, got:", len(ba.BrickSets))
+	tests.Assert(t, len(ba.BrickSets[0].Bricks) == 3,
+		"expected len(ba.BrickSets[0].Bricks) == 3, got:",
+		len(ba.BrickSets[0].Bricks))
+
+	// this will fail because we have no more "arbiter devices"
+	_, err = abplacer.Replace(dsrc, opts, nil, ba.BrickSets[0], 2)
+	tests.Assert(t, err == ErrNoSpace,
+		"expected err == ErrNoSpace, got:", err)
+
+	// this one will work because the free device is not arbiter
+	// and the 1 position is a data brick
+	ba2, err := abplacer.Replace(dsrc, opts, nil, ba.BrickSets[0], 1)
+	tests.Assert(t, err == nil, "expected err == nil, got:", err)
+	tests.Assert(t, len(ba2.BrickSets) == 1,
+		"expected len(ba2.BrickSets) == 1, got:", len(ba2.BrickSets))
+	tests.Assert(t, len(ba2.BrickSets[0].Bricks) == 3,
+		"expected len(ba2.BrickSets[0].Bricks) == 3, got:",
+		len(ba2.BrickSets[0].Bricks))
+}
+
+func init() {
+	SetLogLevel("debug") // DELETEME
```

Comment:
:)

Reply:
I needed it while debugging test failures. I promptly forgot about it. I'll drop it from these changes.